build/build_all.yml

```diff
@@ -0,0 +1,43 @@
+parameters:
+  # GBALLOC_LL_TYPE_VALUES contains values from GBALLOC_LL_TYPE_VALUES of c-pal's CMakeLists.txt.
+  # not all values listed in the CMakeLists need to be listed here - only those for which a completely separate build is done
+  # GBALLOC_LL_TYPE_VALUES is an array of strings
+  - name: GBALLOC_LL_TYPE_VALUES
+    type: object
+    default: ["PASSTHROUGH", "JEMALLOC"]
+
+  # FSANITIZE_TYPE_VALUES are the same as cmakelists "fsanitize" option used to pass (or not) /fsanitize=address to the compiler.
+  - name: FSANITIZE_TYPE_VALUES
+    type: object
+    default: ["ON", "OFF"]
+
+  # ARCH_TYPE_VALUES exists merely for example for future extension capabilities
+  # it provides the "walked path" of how to extend the pipeline with a completely different orthogonal parameter that can take an array of string of values
+  - name: ARCH_TYPE_VALUES
+    type: object
+    #default: ["x64", "x86"]
+    # TODO: for now only x64 is supported, seems that we have some errors on x86
+    default: ["x64"]
+
+  # BUILD_CONFIGURATION_TYPE_VALUES is a list of build configurations to be used in the pipeline
+  - name: BUILD_CONFIGURATION_TYPE_VALUES
+    type: object
+    default: ["Debug", "RelWithDebInfo"]
+
+jobs:
+- ${{ each GBALLOC_LL_TYPE in parameters.GBALLOC_LL_TYPE_VALUES }}:
+  - ${{ each FSANITIZE_TYPE in parameters.FSANITIZE_TYPE_VALUES }}:
+    - ${{ each ARCH_TYPE in parameters.ARCH_TYPE_VALUES }}:
+      - ${{ each BUILD_CONFIGURATION_TYPE in parameters.BUILD_CONFIGURATION_TYPE_VALUES }}:
+        # - ${{ if not( and( eq( GBALLOC_LL_TYPE , 'MIMALLOC' ), eq(FSANITIZE_TYPE, 'ON'))) }}: <==this fails to compile, reasons "2 parameters passed to if". The below line is equivalent.
+        # there is no point in running mimalloc and fsanitize, so that combination is excluded from the whole build.
+        - ${{ if or( ne( GBALLOC_LL_TYPE , 'JEMALLOC' ), ne(FSANITIZE_TYPE, 'ON')) }}:
+          - template: gated_jobs/build_and_run_tests.yml
```

Comment:
>build_and_run_tests [](http://example.com/codeflow?start=33&length=19)

I also wonder if this should be a template in c-build-tools for the general build/test pipelines where we don't need anything more advanced with service fabric/int tests/etc.

We already pull in yml templates from c-build-tools so it should work (in theory)

(just saying since it seems you're duplicating all of this right now anyway, but it's probably more of a task for later...)

Reply:
It should absolutely be in c-build-tools. And used in ctest, c-test-runnerswitcher and all.